lib/firmware/fsm_msg_nano.h

```diff
@@ -0,0 +1,318 @@
+        fsm_sendFailure(FailureType_Failure_Other, _("Block data invalid"));
+        layoutHome();
+        return;
+    }
+
+    // Some additional sanity checks now that balance values are known
+    char recipient_address[MAX_NANO_ADDR_SIZE];
+    memset(recipient_address, 0, sizeof(recipient_address));
+
+    char representative_address[MAX_NANO_ADDR_SIZE];
+    strlcpy(representative_address, msg->representative, sizeof(representative_address));
+    if (msg->has_parent_block) {
+        if (!strncmp(representative_address,
+                     msg->parent_block.representative,
+                     sizeof(representative_address))) {
+            // Representative hasn't changed, zero it out
+            memset(representative_address, 0, sizeof(representative_address));
+        }
+    }
+    if (strlen(representative_address) > 0) {
+        nano_truncate_address(coin, representative_address);
+    }
+
+    bool is_send = false;
+    if (bn_is_less(&balance, &parent_balance)) {
+        is_send = true;
+        bn_subtract(&parent_balance, &balance, &balance_delta);
+    } else {
+        bn_subtract(&balance, &parent_balance, &balance_delta);
+    }
+
+    if (bn_is_zero(&balance_delta)) {
+        // Balance can only remain unchanged when the account exists already
+        invalid |= !msg->has_parent_block;
+    } else if (is_send) {
+        // For sends make fill out the link_recipient (or generate it from link bytes)
+        if (msg->has_link_recipient) {
+            strlcpy(recipient_address, msg->link_recipient, sizeof(recipient_address));
+        } else {
+            nano_get_address(link,
+                coin->nanoaddr_prefix, strlen(coin->nanoaddr_prefix),
+                recipient_address, sizeof(recipient_address));
+        }
+        nano_truncate_address(coin, recipient_address);
+    } else {
+        // For receives make sure that the link_hash was specified and that it's
+        invalid |= !msg->has_link_hash;
+        uint8_t link_empty[sizeof(link)];
+        memset(link_empty, 0, sizeof(link_empty));
+        invalid |= !memcmp(link_empty, link, sizeof(link));
+    }
+
+    if (invalid) {
+        fsm_sendFailure(FailureType_Failure_Other, _("Block data invalid"));
+        layoutHome();
+        return;
+    }
+
+    // Determine what type of prompt to display
+    bool needs_confirm = true;
+    bool is_transfer = false;
+    if (strlen(representative_address) > 0) {
+        // always confirm representative change
+    } else if (!is_send) {
+        // don't bother confirming pure receives
+        needs_confirm = false;
+    } else if (msg->link_recipient_n_count > 0 &&
+               !nano_path_mismatched(coin, msg->link_recipient_n,
+                                     msg->link_recipient_n_count)) {
+        is_transfer = true;
+    }
+
+    if (needs_confirm) {
+        if (strlen(representative_address) > 0) {
+            if (!confirm(ButtonRequestType_ButtonRequest_ConfirmOutput,
```

Comment:
Are representatives' addresses static enough to have a lookup table of the top few well-known ones? That'd allow us to put human-readable names on some of them, which I believe users would be more likely to correctly validate on-screen.

Reply:
I guess the more popular ones are static enough.. but then it would have to match that displayed on the actual desktop wallet, and not sure how many wallets actually display a human readable representative address. So if we did include the lookup table, we'd have to display both the human readable label and the actual address.

Reply:
Good point. To get around that, we could have a separate `NanoLookupRepresentative` / `NanoRepresentative` pair of messages that clients could use to probe the firmware for known representatives before signing. Let's hold off on that though, I don't want to make this more complicated than it needs to be.